text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,392 @@
+- Feature Name: `drop_type_destructuring`
+- Start Date: 2024-12-08
+- RFC PR: [rust-lang/rfcs#3738](https://github.com/rust-lang/rfcs/pull/3738)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Rust does not allow destructuring types which implement the `Drop` trait. This means that moving data out of such types is hard and error prone. The rationale is that once fields are moved out, the type's `Drop` implementation cannot run, which can be undesired. This RFC proposes to allow destructuring anyway, in certain situations.
+
+# Motivation
+[motivation]: #motivation
+
+There are real use-cases in which you do want to move out of types that implement `Drop`. Two main types of patterns that often need it are:
+
+- implementing methods like `into_raw_parts` which explicitly hand over their internals and move the burden of dropping those
+- "guard" types which might need to be "disarmed"
+
+In both cases, you want to avoid the regular `Drop` implementation from running at all and get ownership of the fields.
+
+Right now, code that wants to do this needs to use `unsafe`. For example, [`std::io::BufWriter::into_parts()`](https://doc.rust-lang.org/stable/std/io/struct.BufWriter.html#method.into_parts) has to perform the following gymastics using `ManuallyDrop`:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    let mut this = ManuallyDrop::new(self);
+    let buf = mem::take(&mut this.buf);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    // SAFETY: double-drops are prevented by putting `this` in a ManuallyDrop that is never dropped
+    let inner = unsafe { ptr::read(&this.inner) };
+
+    (inner, buf)
+}
```

Comment:
A safer workaround is to use `Option`, [`Option::take`](https://doc.rust-lang.org/std/option/enum.Option.html#method.take), and decomposing a struct further into components, so that you can "hide" the pieces that need to run the Drop impl from a destructuring pattern. But "Option + decomposition just for the sake of workarounds" is also a really bad negative pattern if you didn't actually need to decompose anything or make the value, er, optional, since it makes the code open to incorrectness, as you may actually want to always have the value there and may want to always have certain values together... *except* in the code that wants to perform this trick.

I can't produce a good example right now, unfortunately, because I lost the logs of the conversation where I explained the (goofy) solution to this question in one case.